spock-core/src/main/java/org/spockframework/runtime/ExtensionRunner.java

```diff
@@ -81,9 +81,35 @@ private void doRunAnnotationDrivenExtensions(Iterable<MethodInfo> nodes) {
     }
   }
 
-  @SuppressWarnings("unchecked")
   private void doRunAnnotationDrivenExtensions(NodeInfo<?, ?> node) {
-    for (Annotation ann : node.getAnnotations()) {
+    RepeatedExtensionAnnotations repeatedExtensionAnnotations = node.getAnnotation(RepeatedExtensionAnnotations.class);
+
+    List<Annotation> annotations;
+    if (repeatedExtensionAnnotations == null) {
+      annotations = Arrays.asList(node.getAnnotations());
+    } else {
+      annotations = new ArrayList<>();
+      List<Class<? extends Annotation>> repeatedAnnotations = Arrays.asList(repeatedExtensionAnnotations.value());
+
+      // add all direct annotations except those marked as repeated
+      Arrays.stream(node.getAnnotations())
+        .filter(annotation -> repeatedAnnotations
+          .stream()
+          .noneMatch(repeatedAnnotation -> repeatedAnnotation.isInstance(annotation)))
```

Comment:
Couldn't we just use a LinkedHashSet to enforce uniqueness?

Reply:
No, because it is not about uniqueness.
Imagine a repeatable annotation `@DelayFor5Seconds` and a user applies it twice.
The expectation would be that he has a delay of 10 seconds, but both annotations are equal so a `Set` would throw out all but one.
Maybe the example is not the best one, but I think you get the idea. :-)

Reply:
What I mean is using
```java
      Class<? extends Annotation> annotations = new LinkedHashSet<>(Arrays.asList(node.getAnnotations()));
      for (Class<? extends Annotation> repeatedAnnotation : repeatedAnnotations) {
        annotations.addAll(Arrays.asList(node.getAnnotationsByType(repeatedAnnotation)));
      }
```

Here is a sample 
```groovy
import javax.validation.constraints.Max

import spock.lang.Specification
import spock.lang.Unroll

class Foo extends Specification{
    @Unroll
    def test() {
        given:
        def val = Bar.getDeclaredMethod(name)
        def anno = new LinkedHashSet(Arrays.asList(val.annotations))
        anno.addAll(Arrays.asList(val.getAnnotationsByType(Max)))

        expect:
        anno.size() == 3

        where:
        name << ['getFoo', 'getBar', 'getBlub']
    }
}


class Bar {

    @Max(1L)
    @Max(2L)
    long getFoo() {
        0
    }


    @Max.List([@Max(4L), @Max(5L)])
    long getBar(){
        0
    }

    @Max(6L)
    @Max.List([@Max(7L)])
    long getBlub(){
        0
    }
}
```